extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/spi/MutinyWrapperHandlerService.java

```diff
@@ -0,0 +1,114 @@
+import java.util.concurrent.CompletionStage;
+
+import org.dataloader.BatchLoaderEnvironment;
+import org.eclipse.microprofile.graphql.GraphQLException;
+
+import graphql.execution.DataFetcherResult;
+import graphql.schema.DataFetchingEnvironment;
+import io.smallrye.graphql.SmallRyeGraphQLServerMessages;
+import io.smallrye.graphql.schema.model.Field;
+import io.smallrye.graphql.schema.model.Wrapper;
+import io.smallrye.graphql.spi.WrapperHandlerService;
+import io.smallrye.graphql.spi.datafetcher.AbstractWrapperHandlerService;
+import io.smallrye.graphql.transformation.AbstractDataFetcherException;
+import io.smallrye.mutiny.Uni;
+
+/**
+ * Handle Async calls with Uni from Mutiny
+ */
+public class MutinyWrapperHandlerService extends AbstractWrapperHandlerService {
+
+    private static final List<String> FOR_CLASSES = Arrays.asList(new String[] {
+            Uni.class.getName()
+    });
+
+    @Override
+    public String getName() {
+        return "Mutiny";
+    }
+
+    @Override
+    public List<String> forClasses() {
+        return FOR_CLASSES;
+    }
+
+    @Override
+    public WrapperHandlerService newInstance() {
+        return new MutinyWrapperHandlerService();
+    }
+
+    @Override
+    protected <T> T invokeAndTransform(DataFetchingEnvironment dfe, DataFetcherResult.Builder<Object> resultBuilder,
+            Object[] transformedArguments) throws AbstractDataFetcherException, Exception {
+        Uni<?> uni = reflectionHelper.invoke(transformedArguments);
+
+        return (T) uni
+                .onItemOrFailure().transform((result, throwable) -> {
+
+                    if (throwable != null) {
+                        if (throwable instanceof CompletionException || shouldUnwrap(throwable)) {
+                            throwable = throwable.getCause();
+                        }
+
+                        eventEmitter.fireOnDataFetchError(dfe.getExecutionId().toString(), throwable);
+                        if (throwable instanceof GraphQLException) {
+                            GraphQLException graphQLException = (GraphQLException) throwable;
+                            partialResultHelper.appendPartialResult(resultBuilder, dfe, graphQLException);
+                        } else if (throwable instanceof Exception) {
+                            throw SmallRyeGraphQLServerMessages.msg.dataFetcherException(operation, throwable);
+                        } else if (throwable instanceof Error) {
+                            throw ((Error) throwable);
+                        }
+                    } else {
+                        try {
+                            resultBuilder.data(fieldHelper.transformResponse(result));
+                        } catch (AbstractDataFetcherException te) {
+                            te.appendDataFetcherResult(resultBuilder, dfe);
+                        }
+                    }
+
+                    return resultBuilder.build();
+                }).subscribeAsCompletionStage();
```

Comment:
It would be great to remote this last line and allow the method to return a Uni.

Reply:
It does not work then.

Reply:
We will need to make the caller be aware of Uni. 
Rewrapping async result in another envelope is not efficient.

Reply:
The caller knows about T (and that is Uni in this case). How would the caller change once it knows about Uni ?

Reply:
This is the caller : https://github.com/smallrye/smallrye-graphql/blob/110406e2f8b508ea899edc8b570d610b4f154528/server/implementation/src/main/java/io/smallrye/graphql/execution/datafetcher/PlugableDataFetcher.java#L34